File: others/sntentop/math-skills.git/python/mathskills.py
import sys # This module provides access to some variables used or maintained by the interpreter and to functions that interact strongly with the interpreter.
import math # This module provides access to the mathematical functions defined by the C standard.

def read_data(file_path):  # Define a function that reads data from a file
    with open(file_path, 'r') as file: # Open the file in read mode
        return [int(line.strip()) for line in file.readlines()] #Read numbers from the file, convert them to integers, and return as a list

def calculate_average(data): # Define a function that calculates the average of a list of numbers
    return sum(data) / len(data) # Return the sum of the numbers divided by the count of numbers

def calculate_median(data): # Define a function that calculates the median of a list of numbers
    sorted_data = sorted(data) # Sort the data in ascending order
    n = len(sorted_data) # Get the count of elements in the sorted data
    middle = n // 2 # Calculate the index of the middle element
    if n % 2 == 0: # Check if the count of elements is even
        
        return (sorted_data[middle - 1] + sorted_data[middle]) / 2 # If even number of elements, return the average of the two middle ones
    else: 
        
        return sorted_data[middle] # If odd, return the middle element

def calculate_variance(data, average): # Define a function that calculates the variance of a list of numbers
    return sum((x - average) ** 2 for x in data) / len(data) # Return the sum of squared differences from the average divided by the count of numbers
 
def calculate_std_deviation(variance): # Define a function that calculates the standard deviation from the variance
    return math.sqrt(variance) # Return the square root of the variance

def custom_round(value): # Define a function that rounds a number to the nearest integer
    return math.ceil(value) if value % 1 >= 0.5 else math.floor(value) # Return the ceiling of the number if the decimal part is greater than or equal to 0.5, otherwise return the floor

def main(): # Define the main function
    if len(sys.argv) != 2: # Check if the number of command-line arguments is not equal to 2
        print("Usage: python your_program.py data.txt") # Print a usage message
        return # Exit the program

    file_path = sys.argv[1] # Get the file path from the command-line arguments
    data = read_data(file_path) # Read data from the file

    average = calculate_average(data) # Calculate the average of the data using the sum of the data and the count of numbers
    median = calculate_median(data) # Calculate the median of the data using the sorted data
    variance = calculate_variance(data, average) # Calculate the variance of the data using the average
    std_deviation = calculate_std_deviation(variance) # Calculate the standard deviation from the variance 

    print(f"Average: {custom_round(average)}") # Print the average rounded to the nearest integer 
    print(f"Median: {custom_round(median)}") # Print the median rounded to the nearest integer 
    print(f"Variance: {custom_round(variance)}") # Print the variance rounded to the nearest integer 
    print(f"Standard Deviation: {custom_round(std_deviation)}") # Print the standard deviation rounded to the nearest integer 

if __name__ == "__main__": # Check if the script is being run directly 
    main()  # Call the main function 

File: sntentop1/sntentop/math-skills.git/go/mathskills.go
package main // package name

import (
	"bufio"   // package for reading files and user input from the console
	"fmt"     // package for printing to the console
	"math"    // package for mathematical functions
	"os"      // package for reading files and user input from the console
	"sort"    // package for sorting numbers
	"strconv" // package for converting strings to numbers
)

func readData(filePath string) ([]int, error) { // function to read data from a file
	file, err := os.Open(filePath) // open the file
	if err != nil {                // if there is an error
		return nil, err // return the error
	} // end if
	defer file.Close() // close the file after the function ends

	var data []int                    // create a slice to store the data as integers
	scanner := bufio.NewScanner(file) // create a scanner to read the file line by line
	for scanner.Scan() {              // loop through each line in the file and read it as a string
		num, err := strconv.Atoi(scanner.Text()) // convert the string to an integer and store it in the num variable as an integer
		if err != nil {                          // if there is an error
			return nil, err // return the error and nil slice
		} // end if
		data = append(data, num) // append the integer to the data slice and continue the loop
	}
	if err := scanner.Err(); err != nil { // if there is an error reading the file after the loop ends
		return nil, err // return the error and nil slice
	}

	return data, nil // return the data slice and nil error and end the function
}

// Calculates the average of a list of numbers
func calculateAverage(data []int) float64 { // function to calculate the average of a list of numbers
	sum := 0                   // variable to store the sum of the numbers
	for _, num := range data { // loop through each number in the list
		sum += num // add the number to the sum
	}
	return float64(sum) / float64(len(data)) // return the average of the numbers
}

// Calculates the median of a list of numbers
func calculateMedian(data []int) float64 { // function to calculate the median of a list of numbers
	sort.Ints(data) // sort the numbers in ascending order using the sort package from the standard library (sort.Ints)
	n := len(data)  // get the length of the list of numbers (n) and store it in the variable n as an integer (int) data type
	mid := n / 2    // calculate the middle index of the list of numbers and store it in the variable mid as an integer (int) data type

	if n%2 == 0 { // if the length of the list of numbers is even (n is divisible by 2 with no remainder)
		return float64(data[mid-1]+data[mid]) / 2 // return the average of the two middle numbers as a float64 data type
	}
	return float64(data[mid]) // if the length of the list of numbers is odd, return the middle number as a float64 data type
}

// Calculates the variance of a list of numbers
func calculateVariance(data []int, average float64) float64 { // function to calculate the variance of a list of numbers
	var sumSquaredDiff float64 // variable to store the sum of the squared differences between each number and the average
	for _, num := range data { // loop through each number in the list of numbers
		diff := float64(num) - average // calculate the difference between the number and the average and store it in the variable diff as a float64 data type
		sumSquaredDiff += diff * diff  // add the square of the difference to the sum of the squared differences
	}
	return sumSquaredDiff / float64(len(data)) // return the variance of the list of numbers as a float64 data type
}

// (Check the math.Pow in go)
// Calculates the standard deviation of a list of numbers
func calculateStdDeviation(variance float64) float64 { // function to calculate the standard deviation of a list of numbers
	return math.Sqrt(variance) // return the square root of the variance as a float64 data type
}

// Custom rounding function that rounds 0.5 and higher up to the next integer
func customRound(value float64) int { // function to round a float64 number to the nearest integer
	if value-math.Floor(value) >= 0.5 { // if the decimal part of the number is greater than or equal to 0.5
		return int(math.Ceil(value)) // round the number up to the next integer using the math.Ceil function from the standard library
	}
	return int(math.Floor(value)) // round the number down to the nearest integer using the math.Floor function from the standard library
}

//(may be use math.Round in go)

func main() { // main function to run the program
	if len(os.Args) != 2 { // if the number of command-line arguments is not equal to 2
		fmt.Println("Usage: go run your_program.go data.txt") // print a usage message to the console
		return                                                // end the program and return a non-zero exit status
	}

	filePath := os.Args[1]          // get the file path from the command-line arguments
	data, err := readData(filePath) // read the data from the file using the readData function
	if err != nil {                 // if there is an error reading the file
		fmt.Printf("Error reading file: %v\n", err) // print an error message to the console
		return                                      // end the program and return a non-zero exit status and end the program
	}

	average := calculateAverage(data)               // calculate the average of the data using the calculateAverage function
	median := calculateMedian(data)                 // calculate the median of the data using the calculateMedian function
	variance := calculateVariance(data, average)    // calculate the variance of the data using the calculateVariance function
	stdDeviation := calculateStdDeviation(variance) // calculate the standard deviation of the data using the calculateStdDeviation function

	fmt.Printf("Average: %d\n", customRound(average))                 // print the average to the console using the customRound function
	fmt.Printf("Median: %d\n", customRound(median))                   // print the median to the console using the customRound function
	fmt.Printf("Variance: %d\n", customRound(variance))               // print the variance to the console using the customRound function
	fmt.Printf("Standard Deviation: %d\n", customRound(stdDeviation)) // print the standard deviation to the console using the customRound function
}
